packages/Telephony/Actions/BatchDTMF.cs

```diff
@@ -0,0 +1,109 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Runtime.CompilerServices;
+using System.Text.RegularExpressions;
+using System.Threading;
+using System.Threading.Tasks;
+
+using AdaptiveExpressions.Properties;
+using Microsoft.Bot.Builder.Dialogs;
+using Newtonsoft.Json;
+
+namespace Microsoft.Bot.Components.Telephony.Actions
+{
+    /// <summary>
+    /// Aggregates input until it matches a regex pattern and then stores the result in an output property
+    /// </summary>
+    public class BatchDTMF : Dialog
+    {
+        /// <summary>
+        /// Class identifier.
+        /// </summary>
+        [JsonProperty("$kind")]
+        public const string Kind = "Microsoft.Telephony.BatchDTMF";
+        private const string AggregationDialogMemory = "dialog.aggregation";
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="BatchDTMF"/> class.
+        /// </summary>
+        /// <param name="sourceFilePath">Optional, source file full path.</param>
+        /// <param name="sourceLineNumber">Optional, line number in source file.</param>
+        [JsonConstructor]
+        public BatchDTMF([CallerFilePath] string sourceFilePath = "", [CallerLineNumber] int sourceLineNumber = 0)
+            : base()
+        {
+            // enable instances of this command as debug break point
+            this.RegisterSourceLocation(sourceFilePath, sourceLineNumber);
+        }
+
+        /// <summary>
+        /// Gets or sets the regex pattern to use to decide when the dialog has aggregated the whole message.
+        /// </summary>
+        [JsonProperty("terminationConditionRegexPattern")]
+        public StringExpression TerminationConditionRegexPattern { get; set; }
```

Comment:
Functionality / extensibility: We are locking in to regex here. We've talked about 3 kinds of termination:

1. Regex is fulfilled
2. Character count is met
3. Configurable terminator character

Would these 3 be desirable still? If yes, what's the future / extensibility story? would we ahve a single DTMF dialog for the 3 terminations or would we have an injectable terminator of some sort?

Reply:
Of course they can all be done through regexes, however I'm thinking from a non-developer perspective, where 2. and 3. are muuuch more approachable.

Reply:
Brainstorm: Another approach is we could have the regex one be the coded action and then have 2 more actions in schemas that use the same class but when we load them from json, if they specify 'terminatorCharacter' property, in our converter we translate that to a well known regex that we set in the class.